src/backend/vector/packed_simd.rs

```diff
@@ -0,0 +1,252 @@
+        #[derive(Copy, Clone, Debug)]
+        #[repr(transparent)]
+        pub struct $ty(core::arch::x86_64::__m256i);
+
+        impl IntoBits<core::arch::x86_64::__m256i> for $ty {
+            #[inline]
+            fn into_bits(self) -> core::arch::x86_64::__m256i {
+                self.0
+            }
+        }
+
+        impl IntoBits<$ty> for core::arch::x86_64::__m256i {
+            #[inline]
+            fn into_bits(self) -> $ty {
+                $ty(self)
+            }
+        }
+
+        impl PartialEq for $ty {
+            #[inline]
+            fn eq(&self, rhs: &$ty) -> bool {
+                unsafe {
+                    let m = core::arch::x86_64::_mm256_cmpeq_epi8(self.0, rhs.0);
+                    core::arch::x86_64::_mm256_movemask_epi8(m) == -1
+                }
+            }
+        }
+
+        impl Eq for $ty {}
+
+        impl Add for $ty {
+            type Output = Self;
+
+            #[inline]
+            fn add(self, rhs: $ty) -> Self {
+                unsafe { core::arch::x86_64::$add_intrinsic(self.0, rhs.0).into_bits() }
+            }
+        }
+
+        impl AddAssign for $ty {
+            #[inline]
+            fn add_assign(&mut self, rhs: $ty) {
+                *self = *self + rhs
+            }
+        }
+
+        impl Sub for $ty {
+            type Output = Self;
+
+            #[inline]
+            fn sub(self, rhs: $ty) -> Self {
+                unsafe { core::arch::x86_64::$sub_intrinsic(self.0, rhs.0).into_bits() }
+            }
+        }
+
+        impl BitAnd for $ty {
+            type Output = Self;
+
+            #[inline]
+            fn bitand(self, rhs: $ty) -> Self {
+                unsafe { core::arch::x86_64::_mm256_and_si256(self.0, rhs.0).into_bits() }
+            }
+        }
+
+        impl BitXor for $ty {
+            type Output = Self;
+
+            #[inline]
+            fn bitxor(self, rhs: $ty) -> Self {
+                unsafe { core::arch::x86_64::_mm256_xor_si256(self.0, rhs.0).into_bits() }
+            }
+        }
+
+        impl BitAndAssign for $ty {
+            #[inline]
+            fn bitand_assign(&mut self, rhs: $ty) {
+                *self = *self & rhs;
+            }
+        }
+
+        impl BitXorAssign for $ty {
+            #[inline]
+            fn bitxor_assign(&mut self, rhs: $ty) {
+                *self = *self ^ rhs;
+            }
+        }
+
+        #[allow(dead_code)]
+        impl $ty {
+            #[inline]
+            pub fn shl<const N: i32>(self) -> Self {
+                unsafe { core::arch::x86_64::$shl_intrinsic(self.0, N).into_bits() }
+            }
+
+            #[inline]
+            pub fn shr<const N: i32>(self) -> Self {
+                unsafe { core::arch::x86_64::$shr_intrinsic(self.0, N).into_bits() }
+            }
+
+            #[inline]
+            pub fn extract<const N: i32>(self) -> $lane_ty {
+                unsafe { core::arch::x86_64::$extract_intrinsic(self.0, N) as $lane_ty }
```

Comment:
Again, I'm a noob to the tricks of SIMD. Can you make a comment somewhere about why it's okay to cast unsigned to signed SIMD, do adds/subs/shifts, and then cast back to unsigned at the end? It's probably true—I just manually checked `assert_eq!(x.overflowing_add(y).0, (x as i8).overflowing_add(y as i8).0 as u8)` for all `x,y: u8`. But some justification here would be very useful.

Reply:
Right, makes sense. Can you write that in a comment and maybe give a link to a source if you can find it? I hope it's not too annoying. We just need to make sure that every choice is thoroughly documented. It's high assurance code, after all 😬

Reply:
Yeah, sorry! I didn't do this in the first place because I still wasn't exactly sure where and how you want it added. Looks great now!